test/e2e/source_api_server_v1_test.go

```diff
@@ -0,0 +1,284 @@
+			//                we can add a json matcher to improve it in the future.
+
+			// Run asserts
+			if tc.expected == "" {
+				time.Sleep(10 * time.Second)
+				eventTracker.AssertNot(recordevents.Any())
+			} else {
+				eventTracker.AssertAtLeast(1, recordevents.MatchEvent(
+					cetest.DataContains(tc.expected),
+				))
+			}
+		})
+	}
+}
+
+func TestApiServerSourceV1EventTypes(t *testing.T) {
+	const (
+		sourceName         = "e2e-apiserver-source-eventtypes"
+		serviceAccountName = "event-watcher-sa"
+		roleName           = "event-watcher-r"
+	)
+
+	client := setup(t, true)
+	defer tearDown(client)
+
+	ctx := context.Background()
+
+	// creates ServiceAccount and RoleBinding with a role for reading pods and events
+	r := resources.Role(roleName,
+		resources.WithRuleForRole(&rbacv1.PolicyRule{
+			APIGroups: []string{""},
+			Resources: []string{"events", "pods"},
+			Verbs:     []string{"get", "list", "watch"}}))
+	client.CreateServiceAccountOrFail(serviceAccountName)
+	client.CreateRoleOrFail(r)
+	client.CreateRoleBindingOrFail(
+		serviceAccountName,
+		testlib.RoleKind,
+		roleName,
+		fmt.Sprintf("%s-%s", serviceAccountName, roleName),
+		client.Namespace,
+	)
+
+	// Label namespace so that it creates the default broker.
+	if err := client.LabelNamespace(map[string]string{sugar.InjectionLabelKey: sugar.InjectionEnabledLabelValue}); err != nil {
+		t.Fatalf("Error annotating namespace: %v", err)
+	}
+
+	// Wait for default broker ready.
+	client.WaitForResourceReadyOrFail(sugarresources.DefaultBrokerName, testlib.BrokerTypeMeta)
+
+	// Create the api server source
+	apiServerSource := rttestingv1.NewApiServerSource(
+		sourceName,
+		client.Namespace,
+		rttestingv1.WithApiServerSourceSpec(
+			sourcesv1.ApiServerSourceSpec{
+				Resources: []sourcesv1.APIVersionKindSelector{{
+					APIVersion: "v1",
+					Kind:       "Event",
+				}},
+				EventMode:          "Reference",
+				ServiceAccountName: serviceAccountName,
+				// TODO change sink to be a non-Broker one once we revisit EventType https://github.com/knative/eventing/issues/2750
+			}),
+	)
+	apiServerSource.Spec.Sink = duckv1.Destination{Ref: &duckv1.KReference{APIVersion: "eventing.knative.dev/v1beta1", Kind: "Broker", Name: sugarresources.DefaultBrokerName, Namespace: client.Namespace}}
+
+	client.CreateApiServerSourceV1OrFail(apiServerSource)
+
+	// wait for all test resources to be ready
+	client.WaitForAllTestResourcesReadyOrFail(ctx)
+
+	eventTypes, err := waitForEventTypes(ctx, client, len(sources.ApiServerSourceEventTypes))
+	if err != nil {
+		t.Fatalf("Waiting for EventTypes: %v", err)
+	}
+	expectedCeTypes := sets.NewString(sources.ApiServerSourceEventTypes...)
+	for _, et := range eventTypes {
+		if !expectedCeTypes.Has(et.Spec.Type) {
+			t.Fatalf("Invalid spec.type for ApiServerSource EventType, expected one of: %v, got: %s", sources.ApiServerSourceEventTypes, et.Spec.Type)
+		}
+	}
+}
```

Comment:
we have a bug in the reconciler.. The reconciler is setting in the status the 6 EventTypes. It should only set the three that correspond to the mode. Say you configure the mode to be ref, we should just set the add,update,delete event types that belong to ref. Not the ones from resource... 
Can you create a separate issue to track it and maybe we can fix it after this lands?

Reply:
Created issue https://github.com/knative/eventing/issues/4180. I will fix it after this lands.